non_preemptive_sjf_scheduling.cpp

```diff
@@ -0,0 +1,296 @@
+     * execution
+     * @returns void
+     */
+
+    void printResult() {
+        cout << "Status of all processes post completion is as follows:" << endl;
+
+        cout << std::setw(17) << left << "Process ID" << std::setw(17) << left
+             << "Arrival Time" << std::setw(17) << left << "Burst Time"
+             << std::setw(17) << left << "Completion Time" << std::setw(17)
+             << left << "Turnaround Time" << std::setw(17) << left
+             << "Waiting Time" << endl;
+
+        for (size_t i{}; i < result.size(); i++) {
+            cout << std::setprecision(2) << std::fixed << std::setw(17) << left
+                 << get<0>(result[i]) << std::setw(17) << left
+                 << get<1>(result[i]) << std::setw(17) << left
+                 << get<2>(result[i]) << std::setw(17) << left
+                 << get<3>(result[i]) << std::setw(17) << left
+                 << get<4>(result[i]) << std::setw(17) << left
+                 << get<5>(result[i]) << endl;
+        }
+    }
+};
+
+    /**
+   * @brief Computes the final status of processes after applying non-preemptive SJF scheduling
+   * @tparam S Data type of Process ID
+   * @tparam T Data type of Arrival time
+   * @tparam E Data type of Burst time
+   * @param input A vector of tuples containing Process ID, Arrival time, and Burst time
+   * @returns A vector of tuples containing Process ID, Arrival time, Burst time,
+   *         Completion time, Turnaround time, and Waiting time
+   */
+   
+    template <typename S, typename T, typename E>
+  vector<tuple<S, T, E, double, double, double>> get_final_status(
+      vector<tuple<S, T, E>> input) {
+    
+      // Sort the processes based on Arrival time and then Burst time
+      sort(input.begin(), input.end(), sortcol<S, T, E>);
+
+      // Result vector to hold the final status of each process
+      vector<tuple<S, T, E, double, double, double>> result(input.size());
+      double timeElapsed = 0;
+
+      for (size_t i = 0; i < input.size(); i++) {
+        // Extract Arrival time and Burst time
+          T arrival = get<1>(input[i]);
+          E burst = get<2>(input[i]);
+
+          // If the CPU is idle, move time to the arrival of the next process
+          if (arrival > timeElapsed) {
+              timeElapsed = arrival;
+          }
+
+          // Update timeElapsed by adding the burst time
+          timeElapsed += burst;
+
+          // Calculate Completion time, Turnaround time, and Waiting time
+          double completion = timeElapsed;
+          double turnaround = completion - arrival;
+          double waiting = turnaround - burst;
+
+          // Store the results in the result vector
+          result[i] = make_tuple(get<0>(input[i]), arrival, burst, completion, turnaround, waiting);
+      }
+    
+      return result;
+  }
+
+/**
+ * @brief Self-test implementations
+ * @returns void
+ */
+static void test() {
+    for (int i{}; i < 1000; i++) {
+        srand(time(nullptr));
```

Comment:
please do not use srand that is a c function, C++ provides the random STL header for a reason